### PR TITLE
Filter out irregular files from backup

### LINKED ISSUE #100: Grain backup fails if /var contains irregular files (sockets, pipes)
Backup failed: Error: Zip process failed. [500]

### PR DESCRIPTION
**Pull request by jparyani**
Fixes #100

### CODE REVIEW

**Comment by kentonv** [comment]
Security issue: Someone could create a file containing a '\n' character in order to inject an arbitrary filename into this list. They could use that to trick us into instructing zip to follow some symbolic link to an arbitrary place in the filesystem.

Luckily, minibox saves us here because it hides all the other user data on the filesystem, so nothing interesting could be exfiltrated. But I think we should also check for filenames containing '\n' and drop them from the list.

**Comment by jparyani** [comment]
Done.

File: shell/server/backup.js [hunk 1]
```diff
@@ -41,6 +41,33 @@ function recursiveRmdirIfExists(dir) {
...
+      walkAndWriteFiles(name, writeStream, replaceRoot, originalRoot);
+    } else if (fileStat.isFile()) {
+      name = replaceRoot + name.slice(originalRoot.length);
+      writeStream.write(name + "\n");
+    }
+  });
+
+  // Include empty directories
+  if (files.length === 0) {
+    dir = replaceRoot + dir.slice(originalRoot.length);
```
Comment (diracdeltas) [inline shell/server/backup.js]: Don't you also need to check that `dir` doesn't contain a newline?

Comment (jparyani) [inline shell/server/backup.js]: I think it's fine. Any call to walkAndWriteFiles except for the first can only occur within the forEach loop, after the check for the new line has already happened. This should ensure `dir` can never have a newline in it, except for on the first call. We enter walkAndWriteFiles with a known path (specifically Path.join(SANDSTORM_GRAINDIR, grainId, "sandbox")), so it should be fine.

That being said, the logic is a bit twisty and I'm not against adding another check just for clarity.

Comment (diracdeltas) [inline shell/server/backup.js]: Ah, makes sense. It wasn't obvious to me that this will only get called with a known entry point.

**Pull request merged**